crates/sui-core/src/scoring_decision.rs

```diff
@@ -0,0 +1,286 @@
+/// Once we have that value, if any authority's absolute deviation / ( MAD / C) < -K then it is deemed
+/// to be a low-value outlier. The values of C and K can be tweaked to change the sensitivity to outliers.
+/// They were chosen based on trial and error to produce reasonable results for score values in the
+/// order of magnitude of 10s - 1000s. If you increase C and decrease K you will see more values
+/// being included as outliers. As the scores get higher in value, outlier sensitivity tends to
+/// decrease using this method.
+///
+/// If we find that we have rated enough validators as low scoring such that we no longer have
+/// quorum with the remaining validators, then we either need to update this method's parameters,
+/// our general approach to finding outliers, or our network is in a bad state. If we need to update
+/// this code, we let it detect this and disable itself for safety reasons. If we have a bad network
+/// state then in the interest of making debugging and investigation easier, disabling the scoring
+/// mechanism will likely be helpful.
+pub fn update_low_scoring_authorities(
+    low_scoring_authorities: &ArcSwap<HashMap<AuthorityName, u64>>,
+    committee: Committee,
+    reputation_scores: ReputationScores,
+    metrics: Option<&AuthorityMetrics>,
+) {
+    if !reputation_scores.final_of_schedule {
+        return;
+    }
+    let mut score_list = vec![];
+    for val in reputation_scores.scores_per_authority.values() {
+        score_list.push(*val as f64);
+    }
+
+    let median_value = median(&score_list);
+    let mut deviations = vec![];
+    let mut abs_deviations = vec![];
+    for (i, _) in score_list.clone().iter().enumerate() {
+        deviations.push(score_list[i] - median_value);
+        abs_deviations.push((score_list[i] - median_value).abs());
+    }
+
+    // adjusted median absolute deviation
+    let mad = median(&abs_deviations) / 0.7;
+    let mut low_scoring = vec![];
+    let mut rest = vec![];
+    for (i, (a, _)) in reputation_scores.scores_per_authority.iter().enumerate() {
+        let temp = deviations[i] / mad;
+        if temp < -2.4 {
+            low_scoring.push(a);
+        } else {
+            rest.push(AuthorityName::from(a));
+        }
+    }
+
+    // make sure the rest have at least quorum
+    let remaining_stake = rest.iter().map(|a| committee.weight(a)).sum::<u64>();
+    let quorum_threshold = committee.threshold::<true>();
+    if remaining_stake < quorum_threshold {
+        warn!(
+            "Too many low reputation-scoring authorities, temporarily disabling scoring mechanism"
+        );
+
+        low_scoring_authorities.swap(Arc::new(HashMap::new()));
+        return;
+    }
+
+    let mut new_inner = HashMap::new();
+    for authority in low_scoring {
+        new_inner.insert(
+            AuthorityName::from(authority),
+            *reputation_scores
+                .scores_per_authority
+                .get(authority)
+                .unwrap_or(&0),
```

Comment:
Based on the code above it seems impossible to not find the authority in the scores - I would convert this to an `expect` instead.